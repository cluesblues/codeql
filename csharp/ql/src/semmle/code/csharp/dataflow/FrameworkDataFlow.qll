/**
 * Provides classes and predicates for definining data-flow through frameworks.
 */

import csharp
private import internal.FrameworkDataFlowImpl as Impl
private import internal.FrameworkDataFlowSpecific::Private
private import internal.DataFlowPublic

class FrameworkCallable = Impl::FrameworkCallable;

/** An unbound method. */
class FrameworkMethod extends FrameworkCallable, Method { }

class ContentList = Impl::ContentList;

/** Provides predicates for constructing content lists. */
module ContentList {
  import Impl::ContentList

  /** Gets the singleton "element content" content list. */
  ContentList element() { result = singleton(any(ElementContent c)) }

  /** Gets a singleton property content stack. */
  ContentList property(Property p) {
    result = singleton(any(PropertyContent c | c.getProperty() = p.getSourceDeclaration()))
  }

  /** Gets a singleton field content stack. */
  ContentList field(Field f) {
    result = singleton(any(FieldContent c | c.getField() = f.getSourceDeclaration()))
  }
}

class SummaryInput = Impl::SummaryInput;

/** Provides predicates for constructing flow-summary input specifications */
module SummaryInput {
  private import semmle.code.csharp.frameworks.system.Collections

  /**
   * Gets an input specification that specifies the qualifier in a call as
   * the input.
   */
  SummaryInput qualifier() { result = TQualifierSummaryInput() }

  /**
   * Gets an input specification that specifies the `i`th argument in a call as
   * the input.
   */
  SummaryInput argument(int i) { result = TArgumentSummaryInput(i) }

  private predicate isCollectionType(ValueOrRefType t) {
    t.getABaseType*() instanceof SystemCollectionsIEnumerableInterface and
    not t instanceof StringType
  }

  /**
   * Gets an input specification that specifies the `i`th argument as the input.
   *
   * `inputStack` is either empty or a singleton element content stack, depending
   * on whether the type of the `i`th parameter of `c` is a collection type.
   */
  SummaryInput argument(FrameworkCallable c, int i, ContentList inputStack) {
    result = argument(i) and
    exists(Parameter p |
      p = c.getParameter(i) and
      if isCollectionType(p.getType())
      then inputStack = ContentList::element()
      else inputStack = ContentList::empty()
    )
  }

  /**
   * Gets an input specification that specifies output from the delegate at
   * argument `i` as the input.
   */
  SummaryInput delegate(int i) { result = TDelegateSummaryInput(i) }

  /**
   * Gets an input specification that specifies output from the delegate at
   * argument `i` as the input.
   *
   * `c` must be a compatible callable, that is, a callable where the `i`th
   * parameter is a delegate.
   */
  SummaryInput delegate(FrameworkCallable c, int i) {
    result = delegate(i) and
    hasDelegateArgumentPosition(c, i)
  }
}

class SummaryOutput = Impl::SummaryOutput;

/** Provides predicates for constructing flow-summary output specifications. */
module SummaryOutput {
  /**
   * Gets an output specification that specifies the qualifier in a call as
   * the output.
   */
  SummaryOutput qualifier() { result = TQualifierSummaryOutput() }

  /**
   * Gets an output specification that specifies the return value from a call as
   * the output.
   */
  SummaryOutput return() { result = TReturnSummaryOutput() }

  /**
   * Gets an output specification that specifies the `i`th argument in a call as
   * the output.
   */
  SummaryOutput argument(int i) { result = TArgumentSummaryOutput(i) }

  /**
   * Gets an output specification that specifies parameter `j` of the delegate at
   * argument `i` as the output.
   */
  SummaryOutput delegate(FrameworkCallable callable, int i, int j) {
    result = TDelegateSummaryOutput(i, j) and
    hasDelegateArgumentPosition2(callable, i, j)
  }
}

class FlowSource = Impl::FlowSource;

/** Provides predicates for constructing flow-source specifications. */
module FlowSource {
  class Range = Impl::FlowSource::Range;

  private class RemoteFlowSource extends Range {
    RemoteFlowSource() { this = "remote" }
  }

  FlowSource remote() { result = any(RemoteFlowSource s).toFlowSource() }

  private class LocalFlowSource extends Range {
    LocalFlowSource() { this = "local" }
  }

  FlowSource local() { result = any(LocalFlowSource s).toFlowSource() }

  private class StoredFlowSource extends Range {
    StoredFlowSource() { this = "stored" }
  }

  FlowSource stored() { result = any(StoredFlowSource s).toFlowSource() }
}

class FlowSink = Impl::FlowSink;

/** Provides predicates for constructing flow-sink specifications. */
module FlowSink {
  class Range = Impl::FlowSink::Range;

  private class RemoteFlowSink extends Range {
    RemoteFlowSink() { this = "remote" }
  }

  FlowSink remote() { result = any(RemoteFlowSink s).toFlowSink() }

  private class HtmlFlowSink extends Range {
    HtmlFlowSink() { this = "html" }

    override predicate isSubsetOf(FlowSink fs) { fs = remote() }
  }

  FlowSink html() { result = any(HtmlFlowSink s).toFlowSink() }

  private class EmailFlowSink extends Range {
    EmailFlowSink() { this = "email" }

    override predicate isSubsetOf(FlowSink fs) { fs = remote() }
  }

  FlowSink email() { result = any(EmailFlowSink s).toFlowSink() }

  private class SqlFlowSink extends Range {
    SqlFlowSink() { this = "sql" }
  }

  FlowSink sql() { result = any(SqlFlowSink s).toFlowSink() }

  private class CommandFlowSink extends Range {
    CommandFlowSink() { this = "command" }
  }

  FlowSink command() { result = any(CommandFlowSink s).toFlowSink() }
}

class FrameworkDataFlow = Impl::FrameworkDataFlow;

module Examples {
  private import semmle.code.csharp.frameworks.system.Text

  /** Data flow for `System.Text.StringBuilder`. */
  class StringBuilderDataFlow extends FrameworkDataFlow {
    StringBuilderDataFlow() { this = "System.Text.StringBuilder" }

    private SystemTextStringBuilderClass getClass() { any() }

    private predicate constructorFlow(
      Constructor c, SummaryInput input, ContentList inputAp, SummaryOutput output,
      ContentList outputAp
    ) {
      c = this.getClass().getAMember() and
      c.getParameter(0).getType() instanceof StringType and
      input = TArgumentSummaryInput(0) and
      inputAp = ContentList::empty() and
      output = TReturnSummaryOutput() and
      outputAp = ContentList::element()
    }

    private predicate methodFlow(
      FrameworkMethod m, SummaryInput input, ContentList inputAp, SummaryOutput output,
      ContentList outputAp, boolean preservesValue
    ) {
      exists(string name | m = this.getClass().getAMethod(name) |
        name = "ToString" and
        input = TQualifierSummaryInput() and
        inputAp = ContentList::element() and
        output = TReturnSummaryOutput() and
        outputAp = ContentList::empty() and
        preservesValue = false
        or
        exists(int i, Type t |
          name.regexpMatch("Append(Format|Line)?") and
          t = m.getParameter(i).getType() and
          input = TArgumentSummaryInput(i) and
          inputAp = ContentList::empty() and
          output = [TQualifierSummaryOutput().(TSummaryOutput), TReturnSummaryOutput()] and
          outputAp = ContentList::element() and
          preservesValue = true
        |
          t instanceof StringType or
          t instanceof ObjectType
        )
      )
    }

    override predicate hasSummary(
      FrameworkCallable c, SummaryInput input, ContentList inputStack, SummaryOutput output,
      ContentList outputStack, boolean preservesValue
    ) {
      (
        this.constructorFlow(c, input, inputStack, output, outputStack) and
        preservesValue = true
        or
        this.methodFlow(c, input, inputStack, output, outputStack, preservesValue)
      )
    }

    override predicate clearsContent(FrameworkCallable c, SummaryInput input, Content content) {
      c = this.getClass().getAMethod("Clear") and
      input = TQualifierSummaryInput() and
      content instanceof ElementContent
    }
  }

  private import semmle.code.csharp.frameworks.System

  /** Data flow for `System.Lazy`. */
  class LazyDataFlow extends FrameworkDataFlow {
    LazyDataFlow() { this = "System.Lazy" }

    private SystemLazyClass getClass() { any() }

    override predicate hasSummary(
      FrameworkCallable c, SummaryInput input, ContentList inputStack, SummaryOutput output,
      ContentList outputStack, boolean preservesValue
    ) {
      preservesValue = true and
      exists(SystemFuncDelegateType t, int i | t.getNumberOfTypeParameters() = 1 |
        c.(Constructor).getDeclaringType() = this.getClass() and
        c.getParameter(i).getType().getSourceDeclaration() = t and
        input = SummaryInput::delegate(c, i) and
        inputStack = ContentList::empty() and
        output = TReturnSummaryOutput() and
        outputStack = ContentList::property(this.getClass().getValueProperty())
      )
      or
      preservesValue = false and
      c = this.getClass().getValueProperty().getGetter() and
      input = TQualifierSummaryInput() and
      inputStack = ContentList::empty() and
      output = TReturnSummaryOutput() and
      outputStack = ContentList::empty()
    }
  }

  private import semmle.code.csharp.frameworks.system.Data
  private import semmle.code.csharp.frameworks.system.data.SqlClient

  class SqlClientDataFlow extends FrameworkDataFlow {
    SqlClientDataFlow() { this = "SqlClientDataFlow" }

    override predicate hasSink(Node n, FlowSink sink) {
      sink = FlowSink::sql() and
      (
        exists(Property p, SystemDataIDbCommandInterface i, Property text |
          text = i.getCommandTextProperty() and
          p.overridesOrImplementsOrEquals(text) and
          n.(ParameterNode).getParameter() = p.getSetter().getAParameter()
        )
        or
        exists(InstanceConstructor ic | ic = n.asExpr().(ObjectCreation).getTarget() |
          ic.getDeclaringType().getABaseType*() instanceof SystemDataIDbCommandInterface and
          ic.getParameter(0).getType() instanceof StringType
        )
      )
    }
  }
}
