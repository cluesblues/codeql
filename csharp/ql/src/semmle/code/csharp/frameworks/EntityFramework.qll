/**
 * Classes modelling EntityFramework and EntityFrameworkCore.
 */

import csharp
private import DataFlow
private import semmle.code.csharp.frameworks.system.data.Entity
private import semmle.code.csharp.frameworks.system.collections.Generic
private import semmle.code.csharp.frameworks.Sql
private import semmle.code.csharp.dataflow.LibraryTypeDataFlow

module DataAnnotations {
  class NotMappedAttribute extends Attribute {
    NotMappedAttribute() {
      this
          .getType()
          .hasQualifiedName("System.ComponentModel.DataAnnotations.Schema.NotMappedAttribute")
    }
  }
}

module EntityFramework {
  /** An EF6 or EFCore namespace. */
  class EFNamespace extends Namespace {
    EFNamespace() {
      this.getQualifiedName() = "Microsoft.EntityFrameworkCore"
      or
      this.getQualifiedName() = "System.Data.Entity"
    }
  }

  /** A taint source where the data has come from a mapped property stored in the database. */
  class StoredFlowSource extends DataFlow::Node {
    StoredFlowSource() {
      this.asExpr() = any(PropertyRead read | read.getTarget() instanceof MappedProperty)
    }
  }

  private class EFClass extends Class {
    EFClass() { this.getDeclaringNamespace() instanceof EFNamespace }
  }

  /** The class `Microsoft.EntityFrameworkCore.DbContext` or `System.Data.Entity.DbContext`. */
  class DbContext extends EFClass {
    DbContext() { this.getName() = "DbContext" }

    Method getAFindMethod() {
      result = this.getAMethod("Find")
      or
      result = this.getAMethod("FindAsync")
    }

    Method getAnUpdateMethod() { result = this.getAMethod("Update") }
  }

  /** The class `Microsoft.EntityFrameworkCore.DbSet<>` or `System.Data.Entity.DbSet<>`. */
  class DbSet extends EFClass, UnboundGenericClass {
    DbSet() { this.getName() = "DbSet<>" }
  }

  /** The class `Microsoft.EntityFrameworkCore.DbQuery<>` or `System.Data.Entity.DbQuery<>`. */
  class DbQuery extends EFClass, UnboundGenericClass {
    DbQuery() { this.hasName("DbQuery<>") }
  }

  /** A generic type or method that takes a mapped type as its type argument. */
  private predicate usesMappedType(UnboundGeneric g) {
    g instanceof DbSet
    or
    g instanceof DbQuery
    or
    exists(DbContext db |
      g = db.getAnUpdateMethod()
      or
      g = db.getAFindMethod()
    )
  }

  /** A type that is mapped to database table, or used as a query. */
  class MappedType extends ValueOrRefType {
    MappedType() {
      not this instanceof ObjectType and
      not this instanceof StringType and
      not this instanceof ValueType and
      (
        exists(UnboundGeneric g | usesMappedType(g) |
          this = g.getAConstructedGeneric().getATypeArgument()
        )
        or
        this.getASubType() instanceof MappedType
      )
    }
  }

  /** A property that is potentially stored and retrieved from a database. */
  class MappedProperty extends Property {
    MappedProperty() {
      this = any(MappedType t).getAMember() and
      this.isPublic() and
      not this.getAnAttribute() instanceof DataAnnotations::NotMappedAttribute
    }
  }

  /** The struct `Microsoft.EntityFrameworkCore.RawSqlString`. */
  class RawSqlStringStruct extends Struct, LibraryTypeDataFlow {
    RawSqlStringStruct() { this.getQualifiedName() = "Microsoft.EntityFrameworkCore.RawSqlString" }

    override predicate callableFlow(
      CallableFlowSource source, CallableFlowSink sink, SourceDeclarationCallable c,
      boolean preservesValue
    ) {
      c = this.getAConstructor() and
      source.(CallableFlowSourceArg).getArgumentIndex() = 0 and
      sink instanceof CallableFlowSinkReturn and
      preservesValue = true
      or
      c = this.getAConversionTo() and
      source.(CallableFlowSourceArg).getArgumentIndex() = 0 and
      sink instanceof CallableFlowSinkReturn and
      preservesValue = true
    }

    ConversionOperator getAConversionTo() {
      result = this.getAMember() and
      result.getTargetType() instanceof RawSqlStringStruct and
      result.getSourceType() instanceof StringType
    }
  }

  /**
   * A parameter that accepts raw SQL. Parameters of type `System.FormattableString`
   * are not included as they are not vulnerable to SQL injection.
   */
  private class SqlParameter extends Parameter {
    SqlParameter() {
      this.getType() instanceof StringType and
      (
        exists(Callable c | this = c.getParameter(0) | c.getName().matches("%Sql"))
        or
        this.getName() = "sql"
      ) and
      this.getCallable().getDeclaringType().getDeclaringNamespace().getParentNamespace*() instanceof
        EFNamespace
      or
      this.getType() instanceof RawSqlStringStruct
      or
      this = any(RawSqlStringStruct s).getAConstructor().getAParameter()
      or
      this = any(RawSqlStringStruct s).getAConversionTo().getAParameter()
    }
  }

  /** A call to a method in EntityFrameworkCore that executes SQL. */
  class EntityFrameworkCoreSqlSink extends SqlExpr, Call {
    SqlParameter sqlParam;

    EntityFrameworkCoreSqlSink() { this.getTarget().getAParameter() = sqlParam }

    override Expr getSql() { result = this.getArgumentForParameter(sqlParam) }
  }

  /** A call to `System.Data.Entity.DbSet.SqlQuery`. */
  class SystemDataEntityDbSetSqlExpr extends SqlExpr, MethodCall {
    SystemDataEntityDbSetSqlExpr() {
      this.getTarget() = any(SystemDataEntity::DbSet dbSet).getSqlQueryMethod()
    }

    override Expr getSql() { result = this.getArgumentForName("sql") }
  }

  /** A call to a method in `System.Data.Entity.Database` that executes SQL. */
  class SystemDataEntityDatabaseSqlExpr extends SqlExpr, MethodCall {
    SystemDataEntityDatabaseSqlExpr() {
      exists(SystemDataEntity::Database db |
        this.getTarget() = db.getSqlQueryMethod() or
        this.getTarget() = db.getExecuteSqlCommandMethod() or
        this.getTarget() = db.getExecuteSqlCommandAsyncMethod()
      )
    }

    override Expr getSql() { result = this.getArgumentForName("sql") }
  }

  /**
   * Custom flow through `StringValues` library class
   */
  class DbContextFlow extends LibraryTypeDataFlow, Class {
    DbContextFlow() { this.getBaseClass*().getSourceDeclaration() instanceof DbContext }

    Property getADbSetProperty(ValueOrRefType elementType) {
      exists(ConstructedClass c |
        result.getType() = c and
        c.getSourceDeclaration() instanceof DbSet and
        elementType = c.getTypeArgument(0) and
        this.hasMember(any(Property p | result = p.getSourceDeclaration()))
      )
    }

    predicate stepFwd(Content c1, Type t1, Content c2, Type t2) {
      exists(Property p1 |
        p1 = this.getADbSetProperty(t2) and
        c1.(PropertyContent).getProperty() = p1 and
        t1 = p1.getType() and
        c2 instanceof ElementContent
      )
      or
      stepFwd(_, _, c1, t1) and
      (
        exists(Property p |
          p.getDeclaringType() = t1 and
          c2.(PropertyContent).getProperty() = p and
          t2 = p.getType()
        )
        or
        exists(ConstructedInterface ci |
          t1.(ValueOrRefType).getABaseType*() = ci and
          not t1 instanceof StringType and
          ci.getSourceDeclaration() instanceof SystemCollectionsGenericIEnumerableTInterface and
          c2 instanceof ElementContent and
          t2 = ci.getTypeArgument(0)
        )
      )
    }

    predicate source(PropertyContent head, AccessPath tail, Property p) {
      exists(ValueOrRefType elementType |
        head.(PropertyContent).getProperty() = this.getADbSetProperty(elementType) and
        p = elementType.getAProperty().getSourceDeclaration() and
        tail = AccessPath::properties(p)
      )
    }

    override predicate callableFlow(
      CallableFlowSource source, AccessPath sourceAp, CallableFlowSink sink, AccessPath sinkAp,
      SourceDeclarationCallable c, boolean preservesValue
    ) {
      exists(Property elementProp |
        preservesValue = true and
        this.hasMethod(c) and
        c.hasName("SaveChanges") and
        source instanceof CallableFlowSourceQualifier and
        exists(PropertyContent sourceHead, AccessPath sourceTail |
          this.source(sourceHead, sourceTail, elementProp) and
          sourceAp = AccessPath::cons(sourceHead, sourceTail)
        ) and
        exists(Property dbSetProp, ValueOrRefType elementType, AccessPath sinkTail |
          dbSetProp = this.getADbSetProperty(elementType) and
          sink.(CallableFlowSinkJump).getTarget() = dbSetProp.getGetter() and
          requiresAp2(_, _, sinkTail, _, elementProp) and
          sinkTail.getHead().(PropertyContent).getProperty() =
            elementType.getAProperty().getSourceDeclaration() and
          sinkAp = AccessPath::cons(any(ElementContent ec), sinkTail)
        )
      )
    }

    pragma[noinline]
    private predicate wer(Content head, Type t1, AccessPath tail, Type t2) {
      this.stepFwd(head, t1, tail.getHead(), t2)
    }

    private predicate requiresAp2(Content head, Type t1, AccessPath tail, Type t2, Property last) {
      this.wer(head, t1, tail, t2) and
      //not tail.contains(head) and
      (
        exists(ValueOrRefType elementType |
          exists(this.getADbSetProperty(elementType)) and
          last = elementType.getAProperty().getSourceDeclaration() and
          tail = AccessPath::property(last)
        )
        or
        exists(Content head0, AccessPath tail0 |
          this.requiresAp2(head0, t2, tail0, _, last) and
          tail = AccessPath::cons(head0, tail0)
        )
      )
    }

    override predicate requiresAccessPath(Content head, AccessPath tail) {
      this.source(head, tail, _)
      or
      requiresAp2(head, _, tail, _, _)
    }
  }
}
