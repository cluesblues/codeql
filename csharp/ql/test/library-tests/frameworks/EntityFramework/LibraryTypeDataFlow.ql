import semmle.code.csharp.dataflow.LibraryTypeDataFlow
import semmle.code.csharp.frameworks.EntityFramework::EntityFramework

query predicate callableFlow(string callable, string flow, boolean preservesValue) {
  exists(EFLibraryTypeDataFlow x, CallableFlowSource source, CallableFlowSink sink, Callable c |
    callable = c.getQualifiedNameWithTypes() and
    flow = source + " -> " + sink
  |
    x.callableFlow(source, sink, c, preservesValue)
    or
    x.callableFlow(source, AccessPath::empty(), sink, AccessPath::empty(), c, preservesValue)
  )
}

query predicate callableFlowAccessPath(string callable, string flow, boolean preservesValue) {
  exists(
    EFLibraryTypeDataFlow x, CallableFlowSource source, AccessPath sourceAp, CallableFlowSink sink,
    AccessPath sinkAp, Callable c
  |
    x.callableFlow(source, sourceAp, sink, sinkAp, c, preservesValue) and
    callable = c.getQualifiedNameWithTypes() and
    flow = source + " [" + sourceAp + "] -> " + sink + " [" + sinkAp + "]"
  |
    sourceAp.length() > 0
    or
    sinkAp.length() > 0
  )
}
