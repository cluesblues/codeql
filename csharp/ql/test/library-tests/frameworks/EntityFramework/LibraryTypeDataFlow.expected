callableFlow
| Microsoft.EntityFrameworkCore.RawSqlString.RawSqlString(string) | argument 0 -> return | false |
| Microsoft.EntityFrameworkCore.RawSqlString.implicit conversion(string) | argument 0 -> return | false |
callableFlowAccessPath
| Microsoft.EntityFrameworkCore.DbContext.SaveChanges() | qualifier [Persons, [], Id] -> jump to get_Persons [[], Id] | true |
| Microsoft.EntityFrameworkCore.DbContext.SaveChanges() | qualifier [Persons, [], Name] -> jump to get_Persons [[], Name] | true |
